templates/todo/common/auth/py/AzureDeveloperCliCredential.py

```diff
@@ -0,0 +1,120 @@
+
+import six
+
+from azure.core.credentials import AccessToken
+from azure.core.exceptions import ClientAuthenticationError
+
+from azure.identity import CredentialUnavailableError
+from azure.identity._internal.decorators import log_get_token
+
+CLI_NOT_FOUND = "Azure Developer CLI could not be found. Please visit https://aka.ms/azure-dev for installation instructions and then, once installed, authenticate to your Azure account using 'azd login'."
+COMMAND_LINE = "azd auth token --output json --scope {}"
+NOT_LOGGED_IN = "Please run 'azd login' from a command prompt to authenticate before using this credential."
+
+
+class AzureDeveloperCliCredential(object):
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        pass
+
+    def close(self) -> None:
+        """Calling this method is unnecessary."""
+
+    @log_get_token("AzureDeveloperCliCredential")
+    def get_token(self, *scopes: str, **kwargs) -> AccessToken:
+        commandString = ' --scope '.join(scopes)
+        command = COMMAND_LINE.format(commandString)
+        output = _run_command(command)
+        token = parse_token(output)
+        if not token:
+            sanitized_output = sanitize_output(output)
+            raise ClientAuthenticationError(
+                message="Unexpected output from Azure Developer CLI: '{}'. \n".format(sanitized_output))
+
+        return token
+
+
+def parse_token(output):
+    try:
+        token = json.loads(output)
+        dt = datetime.strptime(token["expiresOn"], "%Y-%m-%dT%H:%M:%SZ")
+        if hasattr(dt, "timestamp"):
+            # Python >= 3.3
+            expires_on = dt.timestamp()
+        else:
+            # taken from Python 3.5's datetime.timestamp()
+            expires_on = time.mktime((dt.year, dt.month, dt.day, dt.hour, dt.minute, dt.second, -1, -1, -1))
+
+        return AccessToken(token["token"], int(expires_on))
+    except (KeyError, ValueError):
+        return None
+
+
+def get_safe_working_dir():
+    if sys.platform.startswith("win"):
+        path = os.environ.get("SYSTEMROOT")
+        if not path:
+            raise CredentialUnavailableError(message="Azure Developer CLI credential expects a 'SystemRoot' environment variable")
+        return path
+
+    return "/bin"
+
+
+def sanitize_output(output):
+    """Redact tokens from CLI output to prevent error messages revealing them"""
+    return re.sub(r"\"token\": \"(.*?)(\"|$)", "****", output)
+
+
+def _run_command(command):
+    if sys.platform.startswith("win"):
+        args = ["cmd", "/c", command]
+    else:
+        args = ["/bin/sh", "-c", command]
+    try:
+        working_directory = get_safe_working_dir()
+
+        kwargs = {
+            "stderr": subprocess.PIPE,
+            "cwd": working_directory,
+            "universal_newlines": True,
+            "env": dict(os.environ, NO_COLOR="true"),
+        }
+        if platform.python_version() >= "3.3":
+            kwargs["timeout"] = 10
```

Comment:
Same, we don't need to check for py 3.3 support here.

```suggestion
        kwargs = {
            "stderr": subprocess.PIPE,
            "cwd": working_directory,
            "universal_newlines": True,
            "env": dict(os.environ, NO_COLOR="true"),
            "timeout": 10,
        }
```